File: language/src/main/java/cadenza/types/Type.java
package cadenza.types;

import cadenza.values.Closure;
import cadenza.values.Int;
import com.oracle.truffle.api.CompilerDirectives;
import com.oracle.truffle.api.frame.FrameSlotKind;
import com.oracle.truffle.api.interop.UnsupportedTypeException;

import java.util.Objects;

// eventually move to a more hindley-milner style model with quantifiers, but then we need subsumption, unification, etc.
// also this doesn't presuppose if we're heading towards dependently typed languages or towards haskell right now
public abstract class Type {
  public int getArity() { return 0; }
  public abstract void validate(Object t) throws UnsupportedTypeException; // checks the contract for a given type holds, for runtime argument passing, etc.

  public void match(Type expected) throws TypeError {
    if (!equals(expected)) throw new TypeError("type mismatch", this,expected);
  }

  UnsupportedTypeException unsupported(String msg, Object... objects) {
    return UnsupportedTypeException.create(objects,msg);
  }

  public final FrameSlotKind rep; // used to set the starting frameslotkind
  Type(FrameSlotKind rep) { this.rep = rep; }

  @CompilerDirectives.ValueType
  public static class Arr extends Type {
    private final int arity;
    public final Type argument, result;
    public Arr(Type argument, Type result) {
      super(FrameSlotKind.Object);
      this.argument = argument;
      this.result = result;
      this.arity = result.getArity()+1;
    }

    @Override
    public void validate(Object t) throws UnsupportedTypeException {
      if (!(t instanceof Closure)) throw unsupported("expected closure",t);
      Closure c = (Closure)t;
      if (!(this.argument == c.type))
        throw unsupported(
          "expected closure of type: " + this.toString() + ", but received one of type " + c.type.toString(),
          c
        );
    }
    public int getArity() { return arity; }
    @Override
    public int hashCode() {
      return Objects.hash(argument,result);
    }
    @Override
    public boolean equals(Object o) {
      if (!(o instanceof Arr)) return false;
      Arr that = (Arr)o;
      return argument.equals(that.argument) && result.equals(that.result);
    }
  }

  // IO actions represented ML-style as nullary functions
  @CompilerDirectives.ValueType
  public static class IO extends Type {
    public final Type result;
    public IO(Type result) {
      super(FrameSlotKind.Object); // closure
      this.result = result;
    }

    @Override
    public void validate(Object t) throws UnsupportedTypeException {
      throw unsupported("expected io",t);
    }

    @Override
    public int hashCode() {
      return result.hashCode();
    }

    @Override
    public boolean equals(Object o) {
      if (!(o instanceof IO)) return false;
      return result.equals(((IO)o).result);
    }
  }

  public static Type arr(Type a, Type b) { return new Arr(a,b); }
  public static Type io(Type a) { return new IO(a); }
  public static final Type bool = new Type(FrameSlotKind.Boolean) {
    public void validate(Object t) throws UnsupportedTypeException {
      if (!(t instanceof Boolean)) throw unsupported("expected boolean",t);
    }
  };
  public static final Type object = new Type(FrameSlotKind.Object) {
    public void validate(Object t) {}
  };
  public static final Type unit = new Type(FrameSlotKind.Object) {
    public void validate(Object t) throws UnsupportedTypeException {
      if (t != null) throw unsupported("expected unit", t);
    }
  }; // always null. use byte or something else instead?
  public static final Type nat = new Type(FrameSlotKind.Int) {
    public void validate(Object t) throws UnsupportedTypeException {
      if ( ! ((t instanceof Integer && ((int)t)>=0)
         || (t instanceof Int && ((Int)t).isNatural())))
        throw unsupported("expected nat", t);
    }
  };
  public static final Type action = io(unit);
}

File: language/src/main/kotlin/cadenza/types/Type.kt
package cadenza.types

import cadenza.values.Closure
import cadenza.values.BigInt
import com.oracle.truffle.api.CompilerDirectives
import com.oracle.truffle.api.frame.FrameSlotKind
import com.oracle.truffle.api.interop.UnsupportedTypeException
import com.oracle.truffle.api.nodes.ExplodeLoop

import java.util.Objects

// eventually move to a more hindley-milner style model with quantifiers, but then we need subsumption, unification, etc.
// also this doesn't presuppose if we're heading towards dependently typed languages or towards haskell right now
abstract class Type internal constructor(val rep: FrameSlotKind // used to set the starting frameslotkind
) {
  open val arity: Int
    get() = 0

  @Throws(UnsupportedTypeException::class)
  abstract fun validate(t: Any?)  // checks the contract for a given type holds, for runtime argument passing, etc.

  @Throws(TypeError::class)
  fun match(expected: Type) {
    if (!equals(expected)) throw TypeError("type mismatch", this, expected)
  }

  internal fun unsupported(msg: String, vararg objects: Any?): Nothing {
    throw UnsupportedTypeException.create(objects, msg)
  }

  @CompilerDirectives.ValueType
  class Arr(val argument: Type, val result: Type) : Type(FrameSlotKind.Object) {
    override val arity: Int

    init {
      this.arity = result.arity + 1
    }

    @Throws(UnsupportedTypeException::class)
    override fun validate(t: Any?) {
      if (t !is Closure) unsupported("expected closure", t)
      if (this.argument != t.type)
        unsupported(
          "expected closure of type: " + this.toString() + ", but received one of type " + t.type.toString(),
          t
        )
    }

    override fun hashCode(): Int {
      return Objects.hash(argument, result)
    }

    override fun equals(other: Any?): Boolean {
      return other is Arr && argument == other.argument && result == other.result
    }
  }

  // IO actions represented ML-style as nullary functions
  @CompilerDirectives.ValueType
  class IO(val result: Type)// closure
    : Type(FrameSlotKind.Object) {

    @Throws(UnsupportedTypeException::class)
    override fun validate(t: Any?) {
      unsupported("expected io", t)
    }

    override fun hashCode(): Int {
      return result.hashCode()
    }

    override fun equals(other: Any?): Boolean {
      return other is Arr && result == other.result
    }
  }

  companion object {

    val Bool: Type = object : Type(FrameSlotKind.Boolean) {
      @Throws(UnsupportedTypeException::class)
      override fun validate(t: Any?) {
        if (t !is Boolean) unsupported("expected boolean", t)
      }
    }
    val Obj: Type = object : Type(FrameSlotKind.Object) {
      override fun validate(@Suppress("UNUSED_PARAMETER") t: Any?) {}
    }
    val Unit: Type = object : Type(FrameSlotKind.Object) {
      @Throws(UnsupportedTypeException::class)
      override fun validate(t: Any?) {
        if (kotlin.Unit != t) unsupported("expected unit", t)
      }
    } // always null. use byte or something else instead?
    val Nat: Type = object : Type(FrameSlotKind.Int) {
      @Throws(UnsupportedTypeException::class)
      override fun validate(t: Any?) {
        if (!(t is Int && t >= 0 || t is BigInt && t.isNatural()))
          unsupported("expected nat", t)
      }
    }
    val Action = IO(Unit)

    // assumes validity
    @ExplodeLoop
    fun after(t: Type, n: Int): Type {
      var current = t
      for (i in 0 until n)
        current = (current as Type.Arr).result
      return current
    }
  }

}
